### PR TITLE
update kvstore docs

### PR DESCRIPTION
**Pull request by eric-haibin-lin**
#6000

### CODE REVIEW

**Comment by eric-haibin-lin** [comment]
@zackchase @nswamy @madjam

File: python/mxnet/kvstore.py [hunk 1]
```diff
@@ -379,12 +400,33 @@ def _send_command_to_servers(self, head, body):
...
+    ``dist_sync``: it behaves similarly to ``local`` but exhibits one major difference.
+    With ``dist_sync``, batch-size now means the batch size used on each machine.
+    So if there are n machines and we use batch size b,
+    then ``dist_sync`` behaves like local with batch size n * b.
+
+    ``dist_device_sync``: it is identical to ``dist_sync`` with the difference similar
+    to ``device`` vs ``local``.
+
+    ``dist_async`` performs asynchronous updates.
+    The weight is updated whenever gradients are received from any machine. The update is atomic,
```
Comment (madjam) [inline python/mxnet/kvstore.py]: weight is -> weights are

"update is atomic": You mean gradient updates are serialized?

File: python/mxnet/kvstore.py [hunk 2]
```diff
@@ -379,12 +400,33 @@ def _send_command_to_servers(self, head, body):
...
+
+    ``device``: both gradient aggregation and weight updates are run on GPUs. With this setting,
+    the KVStore also attempts to use GPU peer-to-peer communication,
+    potentially accelerating the communication.
+
+    For distributed training, KVStore also supports a number of types:
+
+    ``dist_sync``: it behaves similarly to ``local`` but exhibits one major difference.
+    With ``dist_sync``, batch-size now means the batch size used on each machine.
+    So if there are n machines and we use batch size b,
```
Comment (madjam) [inline python/mxnet/kvstore.py]: n -> `n` and b -> `b`
n * b -> `n * b`

File: python/mxnet/kvstore.py [hunk 3]
```diff
@@ -224,16 +224,35 @@ def pull(self, key, out=None, priority=0):
...
     def set_optimizer(self, optimizer):
-        """ Registers an optimizer with the store.
+        """ Registers an optimizer with the kvstore.
 
-        When there are multiple machines, this operation (invoked from a worker node)
-        will pack the optimizer and send it to all servers. It returns after
+        For single machine, it updates the local optimizer.
+
+        For multiple machines, if this operation is invoked from a worker node,
+        it will pack the optimizer and send it to all servers. The function returns after
```
Comment (piiswrong) [inline python/mxnet/kvstore.py]: pack->serialize the optimizer with pickle

File: python/mxnet/kvstore.py [hunk 4]
```diff
@@ -379,12 +400,33 @@ def _send_command_to_servers(self, head, body):
 def create(name='local'):
```
Comment (piiswrong) [inline python/mxnet/kvstore.py]: @mli we should consider changing default to device?

File: python/mxnet/kvstore.py [hunk 5]
```diff
@@ -379,12 +400,33 @@ def _send_command_to_servers(self, head, body):
...
+
+    ``local``: all gradients are copied to CPU memory and weights are updated there.
+
+    ``device``: both gradient aggregation and weight updates are run on GPUs. With this setting,
+    the KVStore also attempts to use GPU peer-to-peer communication,
+    potentially accelerating the communication.
+
+    For distributed training, KVStore also supports a number of types:
+
+    ``dist_sync``: it behaves similarly to ``local`` but exhibits one major difference.
```
Comment (piiswrong) [inline python/mxnet/kvstore.py]: but exhibits -> with

File: python/mxnet/kvstore.py [hunk 6]
```diff
@@ -290,19 +309,21 @@ def num_workers(self):
         return size.value
 
     def save_optimizer_states(self, fname):
-        """Saves optimizer (updater) state to file.
+        """Saves the optimizer (updater) state to a file.
+
+        This is often used when checkpointing the current progress of training a module.
```
Comment (zackchase) [inline python/mxnet/kvstore.py]: Do we need a one-line paragraph? Can this be part of the first line?

Also it makes more sense to say we checkpoint the model, not that we checkpoint "the progress of a model" agree/disagree?

File: python/mxnet/kvstore.py [hunk 7]
```diff
@@ -379,12 +400,33 @@ def _send_command_to_servers(self, head, body):
 def create(name='local'):
     """Creates a new KVStore.
 
+    For single machine training, there are two commonly used types:
+
+    ``local``: all gradients are copied to CPU memory and weights are updated there.
```
Comment (zackchase) [inline python/mxnet/kvstore.py]: more succinct to say:

"Copies all gradients to CPU memory and updates weights there."

File: python/mxnet/kvstore.py [hunk 8]
```diff
@@ -379,12 +400,33 @@ def _send_command_to_servers(self, head, body):
...
+
+    ``device``: both gradient aggregation and weight updates are run on GPUs. With this setting,
```
Comment (zackchase) [inline python/mxnet/kvstore.py]: Change to:
"Aggregates gradients and updates weights on GPUs. ..."

File: python/mxnet/kvstore.py (hunk 5 above)
Comment (zackchase) [inline python/mxnet/kvstore.py]: Change to "Behaves similarly to ``local`` but ..."

File: python/mxnet/kvstore.py [hunk 9]
```diff
@@ -379,12 +400,33 @@ def _send_command_to_servers(self, head, body):
...
+    the KVStore also attempts to use GPU peer-to-peer communication,
+    potentially accelerating the communication.
+
+    For distributed training, KVStore also supports a number of types:
+
+    ``dist_sync``: it behaves similarly to ``local`` but exhibits one major difference.
+    With ``dist_sync``, batch-size now means the batch size used on each machine.
+    So if there are n machines and we use batch size b,
+    then ``dist_sync`` behaves like local with batch size n * b.
```
Comment (zackchase) [inline python/mxnet/kvstore.py]: "local" should be in backticks

File: python/mxnet/kvstore.py [hunk 10]
```diff
@@ -379,12 +400,33 @@ def _send_command_to_servers(self, head, body):
...
+
+    ``dist_device_sync``: it is identical to ``dist_sync`` with the difference similar
+    to ``device`` vs ``local``.
+
+    ``dist_async`` performs asynchronous updates.
```
Comment (zackchase) [inline python/mxnet/kvstore.py]: Is there a missing colon after ``dist_async``? Also capitalize "Performs"

File: python/mxnet/kvstore.py [hunk 11]
```diff
@@ -379,12 +400,33 @@ def _send_command_to_servers(self, head, body):
...
+    potentially accelerating the communication.
+
+    For distributed training, KVStore also supports a number of types:
+
+    ``dist_sync``: it behaves similarly to ``local`` but exhibits one major difference.
+    With ``dist_sync``, batch-size now means the batch size used on each machine.
+    So if there are n machines and we use batch size b,
+    then ``dist_sync`` behaves like local with batch size n * b.
+
+    ``dist_device_sync``: it is identical to ``dist_sync`` with the difference similar
```
Comment (zackchase) [inline python/mxnet/kvstore.py]: Change to "Identical to ``dist_sync`` ... "

**Pull request merged**